packages/loader/driver-utils/src/runWithRetry.ts

```diff
@@ -30,40 +30,77 @@ export interface IProgress {
     /**
      * Called whenever api returns cancellable error and the call is going to be retried.
      * Any exception thrown from this call back result in cancellation of operation
-     * and propagation of thrown exception.
      * @param delayInMs - delay before next retry. This value will depend on internal back-off logic,
      * as well as information provided by service (like 429 error asking to wait for some time before retry)
      * @param error - error object returned from the call.
      */
     onRetry?(delayInMs: number, error: any): void;
 }
 
-export async function runWithRetry<T>(
+export interface IProgress2 {
+    /**
+     * Abort signal used to cancel operation
+     * Note that most of the layers do not use this signal yet. We need to change that over time.
+     * Please consult with API documentation / implementation.
+     * Note that  number of layers may not check this signal while holding this request in a queue,
+     * so it may take a while it takes effect. This can be improved in the future.
+     * Layers in question are:
+     *    - driver (RateLimiter)
+     *    - runWithRetry
+     */
+    cancel?: AbortSignal;
+
+    /**
+     * Called whenever api returns cancellable error and the call is going to be retried.
+     * @param delayInMs - delay before next retry. This value will depend on internal back-off logic,
+     * as well as information provided by service (like 429 error asking to wait for some time before retry)
+     * @param error - error object returned from the call.
+     * @returns - true to continue retrying, false to abort.
+     */
+    onRetry?(delayInMs: number, error: unknown): boolean;
```

Comment:
if you go with boolean, then name of function should change to "shouldRetry" or something like that.
But I'd think that abort is just one of the cases that we eventually want to add to the flow, i.e. the goal here is not to cherry-pick one particular reason for failed call and treat it separately but move away from exceptions to return types. I.e. this API should return something similar to DriverError (at least in the future, right now we can start on this path by defining single interface in this type).

Reply:
Basically, RunResult should have status: "aborted", but error should eventually be not unknown, but well-defined set of types. We can start with specific interface (with unique errorType) for abort errors, and keep the rest as unknown, adding more and more types to the list and eventually getting rid of " | unknown" part

Reply:
i'm a bit wary of this. i think we should only have on way to abort, and that is via the abort signal. with this there are know two abort paths with seems redundant.